### PR TITLE
feat!: job/step/task API functions and session log assertions

### PR DESCRIPTION
**Pull request by jusiskin**
### What was the problem/requirement? (What/Why)

There were no high-level APIs for asserting about contents in the CloudWatch session logs uploaded by the worker agent.

### What was the solution? (How)

To better model the AWS Deadline Cloud resource model and APIs, I added classes and functions that allow developers to traverse from `Job` &rarr; `Step` &rarr; `Task` &rarr; `Session` which enables test cases using jobs with multiple steps and tasks. The `Session` class provides a `get_session_log()` method for the common use-case of collecting the session log events into memory for use in a test.

Further, the `Session` class has an even higher-level method (`Session.assert_log_contains(...)`) for asserting that a Python regular expression pattern is found in the concatenated session log in CloudWatch. In addition, this method employs exponential back-off to help account for the eventually-consistent property of logs becoming available in CloudWatch.

The exponential backoff multiplier can be configured (defaults to 300ms) and the number of backoff/retries can also be configured (default to 4). The total back-off is 4.5s using the defaults.

Another common use-case is submitting a job that has a single step and task and looking for a pattern in the task's session log. For this, a new `Job.assert_single_task_log_contains(...)` method was added.

### What is the impact of this change?

Consumers of `deadline-cloud-test-fixtures` now have high-level APIs for traversing from a job through steps, tasks, and sessions, and session logs. Consumer have high-level APIs for common test use-cases to assert that a regex pattern exists in the session logs.

#### Unit Tests

Added unit tests for the new functionality

#### E2E Tests

Installed the modified `deadline-cloud-test-fixtures` into a `deadline-cloud-worker-agent` checkout dev env

1.  Ran existing Linux and Windows E2E tests and confirmed they still pass
2.  Modified [this existing Linux job submission E2E test code](https://github.com/aws-deadline/deadline-cloud-worker-agent/blob/9b924a084217b344b4e845461a071f0639edd641/test/e2e/linux/test_job_submissions.py#L103-L116) to use the new `Job.assert_single_task_log_contains(...)` which makes downstream calls to all of the newly added APIs. The changed code was:

    ```py
    job.assert_single_task_log_contains(
        deadline_client=deadline_client,
        logs_client=boto3.client(
            "logs",
            config=botocore.config.Config(retries={"max_attempts": 10, "mode": "adaptive"}),
        ),
        expected_pattern=rf'I am: {job_run_as_user.user}',
    )
    ```

    The test passed successfully.

### Was this change documented?

No

### Is this a breaking change?

Yes:

1.  in PR feedback iterations, we decided to change the existing `Job.lifecycle_status` attribute from a `str` &rarr; `JobLifecycleStatus` enum (see https://github.com/aws-deadline/deadline-cloud-test-fixtures/pull/150#discussion_r1745823449)
2. It was discovered there was a `TaskStatus.UNKNOWN` enum value that is not part of the AWS Deadline Cloud API so it was removed (see [ref docs](https://docs.aws.amazon.com/deadline-cloud/latest/APIReference/API_GetTask.html#deadlinecloud-GetTask-response-runStatus))

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/deadline_test_fixtures/deadline/resources.py [hunk 1]
```diff
@@ -332,7 +393,17 @@ class Job:
     template: dict
 
     name: str
-    lifecycle_status: str
+    lifecycle_status: Literal[
```
Comment (AWS-Samuel) [inline src/deadline_test_fixtures/deadline/resources.py]: Should we change this to use the `JobLifecycleStatus` enum?

File: src/deadline_test_fixtures/deadline/resources.py [hunk 2]
```diff
@@ -652,11 +831,390 @@ def __str__(self) -> str:  # pragma: no cover
...
+    queue: Queue
+    job: Job
+    id: str
+
+    name: str
+    created_at: datetime.datetime
+    created_by: str
+    lifecycle_status: Literal[
+        "CREATE_COMPLETE", "UPDATE_IN_PROGRESS", "UDATE_FAILED", "UPDATE_SUCCEEDED"
+    ]
```
Comment (AWS-Samuel) [inline src/deadline_test_fixtures/deadline/resources.py]: Typo: `UDATE_FAILED` -> `UPDATE_FAILED`


Likewise, should we change this to use `StepLifecycleStatus`?

File: src/deadline_test_fixtures/deadline/resources.py (hunk 1 above)
Comment (jusiskin) [reply src/deadline_test_fixtures/deadline/resources.py]: Discussed offline and the one concern is that this PR becomes a breaking change. We decided that this was an overall improvement valuable enough to justify the breaking change. Rationale:

1.  Using an enum will help test authors with IDEs that support type hints and code completion. Supporting IDEs will suggest the possible values for the enum making it easier to write tests.
2. We are pre-`1.0` and we believe that there are very few consumers of this package &mdash; likely only consumed by AWS Deadline Cloud repositories.

I've switched to the enum and marked the commit as a breaking change

File: src/deadline_test_fixtures/deadline/resources.py (hunk 2 above)
Comment (jusiskin) [reply src/deadline_test_fixtures/deadline/resources.py]: Switched to `StepLifecycleStatus`

File: src/deadline_test_fixtures/deadline/resources.py [hunk 3]
```diff
@@ -299,8 +302,34 @@ class StrEnum(str, Enum):
...
 
 
+class JobLifecycleStatus(StrEnum):
+    ARCHIVED = "ARCHIVED"
+    CREATE_COMPLETE = "CREATE_COMPLETE"
+    CREATE_FAILED = "CREATE_FAILED"
+    CREATE_IN_PROGRESS = "CREATE_IN_PROGRESS"
+    UPDATE_FAILED = "UPDATE_FAILED"
+    UPDATE_IN_PROGRESS = "UPDATE_IN_PROGRESS"
+    UPDATE_SUCCEEEDED = "UPDATE_SUCCEEDED"
```
Comment (AWS-Samuel) [inline src/deadline_test_fixtures/deadline/resources.py]: nit
```suggestion
    UPDATE_SUCCEEDED = "UPDATE_SUCCEEDED"
```

Comment (jusiskin) [reply src/deadline_test_fixtures/deadline/resources.py]: Good catch, thanks! Fixed

**Review by AWS-Samuel — APPROVED** [review]

**Review by moorec-aws — APPROVED** [review]

**Pull request merged**